fix(upgrade): fix/improve support for lifecycle hooks

gkalpak opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
- Lifecycle hooks are called on the "binding destination" (either the controller instance or the scope depending on the value of `bindToController`).
- The `$onInit()` lifecycle hook is called _after_ the linking phase.
- The `$postLink()` lifecycle hook is not supported.


**What is the new behavior?**
- With the exception of `$onChanges`, all lifecycle hooks are called on the controller instance, regardless of the value of `bindToController` (as happens in ng1).
- The `$onInit()` lifecycle hook is called _before_ the linking phase (as happens in ng1).
- The `$postLink` lifecycle hook is supported.


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
[x] Maybe
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: 

Theoretically, this is a fix. But if an application was relying on the previous (inconsistent with ng1) behavior and timing (which they shouldn't), it might break.

gkalpak commented:
BTW, this is a follow-up to #12875.

chuckjaz commented:
Please update the commit comments to use "fix(upgrade):" as a prefix instead of "fix(UpgradeNg1Component):"

gkalpak committed:
fix(upgrade): fix lifecycle hooks calling context and timing

With the exception of `$onChanges()`, all lifecycle hooks in ng1 are called on
the controller, regardless if it is the binding destination or not (i.e.
regardless of the value of `bindToController`).

This change makes `upgrade` mimic that behavior when calling lifecycle hooks.

Additionally, calling the `$onInit()` hook has been moved before calling the
linking functions, which also mimics the ng1 behavior.

gkalpak committed:
feat(upgrade): add support for the `$postLink` lifecycle hook

gkalpak commented:
Rebased (to account for 9092680) and updated commit message scopes.

vicb commented:
@gkalpak @petebacondarwin what is the status here ?

gkalpak commented:
It needs to be reviewed :smiley:

mgol commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -313,6 +326,12 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
     return controller;
   }
 
+  private callLifecycleHook(method: LifecycleHook, context: IBindingDestination, arg?: any) {
```
Why is this indirection needed?

mgol replied:
Ah, because of the `if`. OK.

mgol commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -871,160 +871,551 @@ export function main() {
+             const element = html(`<div><ng2></ng2></div>`);
+             adapter.bootstrap(element, ['ng1']).ready((ref) => {
+               expect($postLinkSpyA).toHaveBeenCalled();
+               expect($postLinkSpyB).toHaveBeenCalled();
```
In many test cases, like this one, it's not actually verified if the hooks are called in a proper moment during the lifecycle of the component, just that they're called. I feel more thorough tests are needed.

gkalpak replied:
:+1:
[There is a test][1] for the AoT version that at least verifies the relative order of lifecycle hooks (it is not perfect, but it is better than not having it). Maybe we should port it over and also expand the current tests.
<sub>(But this is not directly related to this PR.)</sub>

[1]: https://github.com/angular/angular/blob/223870e38019f60348275f1bd12c8237aba73b56/modules/%40angular/upgrade/test/aot/integration/upgrade_component_spec.ts#L2528

mgol replied:
The order is nice but it still doesn't verify the hooks are called in proper moments (i.e. after all prerequisite operations for calling them have been finished).

But I agree it's not necessarily related to this PR.

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -871,160 +871,551 @@ export function main() {
+                                      bindToController: false,
+                                      controllerAs: '$ctrl',
+                                      controller: function() { this.$onInit = $onInitSpyB; }
+                                    }))
```
My understanding here is that you want to test the two scenarios where `$onInit` is provided on the controller instance and on the controller's prototype. So why does one have `bindToController` set to `true` and the other  one set to `false`?

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -871,160 +871,551 @@ export function main() {
+                                      controller: function($scope: angular.IScope) {
+                                        $scope['$onInit'] = $onInitSpy;
+                                      }
+                                    }))
```
Similar to the test for `$onInit` on controllers, there seems to be two variables here: on the prototype or not and bind to controller or not. If these are independent then perhaps we need four checks?

gkalpak replied:
Yes, that's right: We have two independent variables (`bindToControler: true/false` and `prototype vs instance`). Actually, we have a third one which is `scope vs controller`. So, ideally we need 8 different components tested per hook (&times; 5 hooks &times; 2 (dynamic + static upgrade) = 80 components). I cheated by testing two variables per component to save 50% of components tested.

In an ideal world, where `bindToController` is indeed independent from `prototype vs instance`, this should be OK. In the non-ideal world that we live in, we need the 40 extra components (or double as much faith 😛 )

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -871,160 +871,551 @@ export function main() {
+               ref.dispose();
+             });
+           }));
+        it('should not call `$onDestroy()` on scope', fakeAsync(() => {
```
Missing carriage return here?

petebacondarwin reviewed (approved):
I think this is good enough to merge. I left a few minor comments that can be fixed up in subsequent PRs. In particular I think we should have a test that logs all the hook calls to some array so that we can easily see the order of the hooks.

gkalpak commented:
(I just added a missing newline to fix [comment][1], hence the rebase.)

[1]: https://github.com/angular/angular/pull/13020#discussion_r92996662

chuckjaz commented:
@gkalpak Can you rebase this to the current master to resolve conflicts?

gkalpak commented:
@chuckjaz, rebased.

I added one more fixup commit with the changes I made while resolving conflicts. Let me know if you prefer me to squash it.

petebacondarwin commented:
@gkalpak the extra commit looks good. I think you should squashi it, so that it the merge is as straightforward as possible for the caretaker.

gkalpak commented:
Done.

chuckjaz commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -313,6 +326,12 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
 
+  private callLifecycleHook(method: LifecycleHook, context: IBindingDestination, arg?: any) {
+    if (context && typeof context[method] === 'function') {
+      context[method](arg);
```
This style of invocation breaks Closure advanced optimizations and we try to avoid this. Can you change this to avoid the dynamic invocation? I would prefer the code duplication than using dynamic indexing to avoid it.

gkalpak replied:
@chuckjaz, can you give me some more context? Do the optimizations cause the resulting code to break or does the dynamic invocation prevent the optimizations from being applied?

> I would prefer the code duplication than using dynamic indexing to avoid it.

Do you mean remove the `callLifecycleHook` method altogether and instead "inline" the code wherever the method was called?

petebacondarwin replied:
So it is possible that inside a Closure compiled app the `$onInit`, for example, method might be renamed to something "smaller" like `k`. In this case this function would break the behaviour since it would still be trying to call `$onInit`.

We are not sure if this is the case, and it might be possible to get around it be annotating the function to prevent renaming.

But @chuckjaz is suggesting that it is safer simply to not do this in the first place.

gkalpak replied:
Done.

chuckjaz commented:
@gkalpak Can you fix the formatting break? Ensure the `gulp lint` is clean.

gkalpak commented:
@chuckjaz, I tend to forget about that 😞
Fixed now.
